Add common component conditions

Shatur opened the pull request:
Add two useful common run conditions for components similar to `any_with_component` from #7579. Unlike `any_with_component` these two conditions could be potentially expensive, but super useful if user knows that there is only a few entities with a component.

---

## Changelog

### Added

- Add `any_component_added` and `any_component_changed`.

alice-i-cecile reviewed (changes requested):
> Otherwise this check could be expensive and hold up the executor preventing it from running any systems during the check.

I think this can be more clearly phrased by replacing your warning with the following:

> Run conditions are evaluated on the main thread, blocking any other systems from running.
> This run condition is relatively expensive, as it iterates over every entity with this component.
> As a result, you likely only want to use this run condition when the number of entitities with the component `T` is small.

One more request: can we have the same flavor of run condition, but for `RemovedComponent<T>`?

Shatur commented:
> I think this can be more clearly phrased by replacing your warning with the following:

I like it!

> One more request: can we have the same flavor of run condition, but for RemovedComponent<T>?

I think we switched to events for this: https://github.com/bevyengine/bevy/pull/5680

alice-i-cecile commented:
> I think we switched to events for this: https://github.com/bevyengine/bevy/pull/5680

That PR was effectively non-breaking :) The `RemovedComponents` system param is still around, it's just now double-buffered and no longer quite as bespoke.

james7132 commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -141,4 +143,30 @@ pub mod common_conditions {
+    /// This run condition is relatively expensive, as it iterates over every entity with this component.
+    /// As a result, you likely only want to use this run condition when the number of entitities with the component `T` is small.
+    pub fn any_component_added<T: Component>() -> impl FnMut(Query<(), Added<T>>) -> bool {
+        move |query: Query<(), Added<T>>| !query.is_empty()
```
These potentially do a full world scan in the multithreaded executor, as the only early return is if no archetypes match the query. This is relatively easy performance footgun, same with the other condition below.

Shatur replied:
Yep, this is what I mentioned in the description. I thought it's fine if documented. But I understand the controversy. This is why it wasn't included in https://github.com/bevyengine/bevy/pull/7579.

andresovela commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -141,4 +143,30 @@ pub mod common_conditions {
+    ///
+    /// Run conditions are evaluated on the main thread, blocking any other systems from running.
+    /// This run condition is relatively expensive, as it iterates over every entity with this component.
+    /// As a result, you likely only want to use this run condition when the number of entitities with the component `T` is small.
```
```suggestion
    /// As a result, you likely only want to use this run condition when the number of entities with the component `T` is small.
```

inodentry commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -141,4 +143,30 @@ pub mod common_conditions {
+    /// Generates a [`Condition`](super::Condition)-satisfying closure that returns `true`
+    /// if there are any entities with the removed given component type.
+    pub fn any_component_removed<T: Component>() -> impl FnMut(RemovedComponents<T>) -> bool {
+        move |mut removals: RemovedComponents<T>| !removals.iter().count() != 0
```
```suggestion
        move |mut removals: RemovedComponents<T>| !removals.is_empty()
```
no need to iterate

inodentry commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -141,4 +143,30 @@ pub mod common_conditions {
+    }
+
+    /// Generates a [`Condition`](super::Condition)-satisfying closure that returns `true`
+    /// if there are any entities with the removed given component type.
```
```suggestion
    /// if there are any entities with the given component type removed.
```

inodentry commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -141,4 +143,30 @@ pub mod common_conditions {
+    }
+
+    /// Generates a [`Condition`](super::Condition)-satisfying closure that returns `true`
+    /// if there are any entities with the changed given component type.
```
```suggestion
    /// if the value of the given component type has been changed on any entities.
```

inodentry reviewed (changes requested):
Suggesting better wording for the doc comments!

inodentry commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -141,4 +143,30 @@ pub mod common_conditions {
     }
+
+    /// Generates a [`Condition`](super::Condition)-satisfying closure that returns `true`
+    /// if there are any entities with the added given component type.
```
Whoops, missed one.

```suggestion
    /// if the given component type was added to any entities.
```

Shatur commented:
Thanks!

inodentry commented:
Other than the little things I pointed out above, I approve! Given the warning in the documentation, I think this is worth including in Bevy.

As implemented, if, in the future, change detection gets optimized so that the queries don't have to iterate, this API will just magically become fast, and we can remove the perf warning. :)

inodentry commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -145,4 +147,30 @@ pub mod common_conditions {
+    /// Generates a [`Condition`](super::Condition)-satisfying closure that returns `true`
+    /// if there are any entities with the given component type removed.
+    pub fn any_component_removed<T: Component>() -> impl FnMut(RemovedComponents<T>) -> bool {
+        move |removals: RemovedComponents<T>| !removals.is_empty()
```
Wait. My suggestion was wrong. I did not realize that `RemoveComponents` is now based on events, and therefore the same caveats as my `on_event` run condition apply. The events need to be drained, just checking `is_empty` is not enough. Your old iteration code was correct, and my suggestion is wrong. Please revert it. I'm sorry for misleading you.

Shatur replied:
No problem, reverted!

inodentry replied:
You could also add a comment pointing out that `RemovedComponents` is based on events and therefore events need to be drained (similar to the comment i added for `on_event` in #7579), to avoid confusion and make sure that nobody accidentally messes this up in the future.

Shatur replied:
Agree, added!

Shatur commented:
I moved any_component_removed to a separate PR (https://github.com/bevyengine/bevy/pull/8326) that was merged. Updated this PR with to the latest master just in case someone will be interested to include it.

Shatur commented:
It was decided to not implement it, closing.
